book/tutorials/photon_classifier.py

```diff
@@ -0,0 +1,596 @@
+# reflectance) of the Sentinel-2 pixel.
+#
+# ```{note}
+# Sea ice can move very quickly in minutes, so while we've tried our best to find a
+# Sentinel-2 image that was captured at about the same time as the ICESat-2 track, it is
+# very likely that we will still be misclassifying some of the points below because the
+# image and point clouds are not perfectly aligned.
+# ```
+
+# %%
+# Get first STAC item from collection
+item = item_collection.items[0]
+item
+
+
+# %%
+# Get RGB bands from Sentinel-2 STAC item as an xarray.DataArray
+da_image = stackstac.stack(
+    items=item,
+    assets=["red", "green", "blue"],
+    resolution=60,
+    rescale=False,  # keep original 14-bit scale
+    dtype=np.float32,
+    fill_value=np.float32(np.nan),
+)
+da_image = da_image.squeeze(dim="time")  # drop time dimension so only (band, y, x) left
+da_image
+
+
+# %%
+# Get bounding box of Sentinel-2 image to spatially subset ATL07 geodataframe
+bbox = shapely.geometry.box(*da_image.rio.bounds())  # xmin, ymin, xmax, ymax
+gdf = gdf.to_crs(crs=da_image.crs)  # reproject point cloud to Sentinel-2 image's CRS
+gdf = gdf[gdf.within(other=bbox)].reset_index()  # subset point cloud to Sentinel-2 bbox
+
+# %% [markdown]
+# Let's plot the Sentinel-2 satellite image and ATL07 points! The ATL07 data has a
+# `height_segment_ssh_flag` variable with a binary sea ice (0) / sea surface (1)
+# classification, which we'll plot using two different colors to check that things
+# overlap more or less.
+
+# %%
+fig = pygmt.Figure()
+
+# Plot Sentinel-2 RGB image
+# Convert from 14-bit to 8-bit color scale for PyGMT
+fig.grdimage(grid=((da_image / 2**14) * 2**8).astype(np.uint8), frame=True)
+
+# Plot ATL07 points
+# Sea ice points in blue
+df_ice = gdf[gdf.height_segment_ssh_flag == 0].get_coordinates()
+fig.plot(x=df_ice.x, y=df_ice.y, style="c0.2c", fill="blue", label="Sea ice")
+# Sea surface points in orange
+df_water = gdf[gdf.height_segment_ssh_flag == 1].get_coordinates()
+fig.plot(x=df_water.x, y=df_water.y, style="c0.2c", fill="orange", label="Sea surface")
+fig.legend()
+fig.show()
+
+
+# %% [markdown]
+# Looking good! Notice how the orange (sea surface) coincide with the cracks in the sea
+# ice.
+
+
+# %% [markdown]
+# Next, we'll want to do better than a binary 0/1 or ice/water classification, and pick
+# up different shades of gray (white thick ice, gray thin ice, black water). Let's
+# use the X and Y coordinates from the point cloud to pick up surface reflectance values
+# from the Sentinel-2 image.
+#
+# PyGMT's [`grdtrack`](https://www.pygmt.org/v0.12.0/api/generated/pygmt.grdtrack.html)
+# function can be used to do this X/Y sampling from a 1-band grid.
+
+# %%
+df_red = pygmt.grdtrack(
+    grid=da_image.sel(band="red").compute(),  # Choose only the Red band
+    points=gdf.get_coordinates(),  # x/y coordinates from ATL07
+    newcolname="red_band_value",
+    interpolation="n",  # nearest neighbour
+)
```

Comment:
The first time I ran this on cryocloud I got a traceback: 

```pytb
RuntimeError: Error opening 'https://sentinel-cogs.s3.us-west-2.amazonaws.com/sentinel-s2-l2a-cogs/2/C/ND/2019/10/S2B_2CND_20191031_0_L2A/B04.tif': RasterioIOError('No driver registered.')
```
But re-running succeeded... might be an intermittent thing

Reply:
Yes, I have to re-run this sometimes too. Maybe there's a short-lived token or something.